### ISSUE TITLE
Clarifying Questions

### ISSUE THREAD

**Issue by jdwieland8282** [w3c/web-advertising#80]
Hi @benjaminsavage, these questions stem from a conversation between @michaelkleber @tomkershaw1, and a few others here (https://github.com/jkarlin/floc/issues/20).  

They come from the perspective of an ad exchange, who seeks to maximize publisher revenue. 

First off, thanks for preparing the proposal, the specific use case we are attempting to solve was defined by Michael in issue 20 as such:

_1b. "Lookalike segments" where you have some group of people (maybe a retargeting segment from 1a) and you want to construct a larger group of people who are kind of similar to that group. It would be fine to advertise to these folks using a private mechanism like TURTLEDOVE; the problem is finding them in the first place._ -MK

_We assume that a publisher website is able to collect first-party data about its own users._
1. How much FPD would a publisher need to pass this threshold? 
2. If I own swimming.com, and I have FPD about my visitors, my lookalike audiences will attempt to find other people interested in Swimming?  Is that the idea?

_From this first-party data, we presume the publisher is able to train a neural network which can be used to produce an “embedding vector” for each user_
1. As the owner of swimming.com, it sounds like I need to be prepared to produce the embedding vector myself, or hire someone to do it?

_[Step 2] When this person later visits the advertiser’s website in this browser, some 3rd party JavaScript provided by the publisher should be invoked to flush this data to a server-side aggregation service._
1. as the owner of swimming.com do I need to convince advertisers like TYR & Roka to deploy my JS for this to work?

_[step 4] The publisher could then request this aggregated vector via some kind of API. The server-side aggregation service would add some small amount of noise to the aggregation to ensure a target ε-differential privacy. With the exception of supporting vector primitives, this is exactly what is described in Chrome’s current explainer._
1. after this occurs, as a publisher what specifically do I have?  Do I have a vector id, or do I know the user browsing my site is interested in x, y, or z?

**Comment by benjaminsavage** [comment]
Hi there @jdwieland8282 

> How much FPD would a publisher need to pass this threshold?

There is no "threshold". You would simply need enough information to be capable of sensibly assigning an "embedding vector" to each browser instance. If you have very little information, the quality of the data will be poor and the "embedding vectors" assigned will likely carry very little meaning or value.

Embeddings must be trained on something. The example given in the proposal was ad clicks. One could attempt to train an embedding on ad clicks. The training examples would be something like: "This user clicked on this ad", and "This user did NOT click on this ad". You could train an embedding with a sufficiently large number of such data points. 

This isn't the only approach. Perhaps a publisher could train an embedding on non-ads interactions. For example, Tik-Tok could train an embedding based on what videos people clicked "like" on, or Pintrest could train an embedding based on what content people "Pin". All you really need is a large number of positive and negative training examples that can be used to train a system that learns which users are "similar" and which users are "dis-similar". So in my "Tik-Tok" example, I like videos of people snowboarding and playing minecraft. Tik-Tok doubtless has many similar users who like similar types of content (as evidenced by the videos they like). An embedding would automatically figure that out and assign me some point in N-space that is similar to the points in N-space of other people who like snowboarding and minecraft videos. 

> If I own swimming.com, and I have FPD about my visitors, my lookalike audiences will attempt to find other people interested in Swimming? Is that the idea?

No, I don't think this is the idea. I assume you're using "swimming.com" as an example of a publisher who might know what people are interested in by virtue of having a specific type of content on their website. This isn't what I had in mind. As explained above, I was more thinking of publishers that had enough first-party data about their userbase to be able to understand which users of their site were similar to other users of their site, and thereby train an embedding that would assign to each of their users an "embedding vector" that somehow represents this concept of similarity.

What type of first-party-data might swimming.com have? Firstly, it's a parked domain, so nothing =). However, we might imagine it as some kind of information hub full of swimming related information. What kind of first-party-data might such a website have? Well, perhaps there is a login (perhaps there is not). Perhaps people read articles? Perhaps they comment on articles? I'm not really sure what to imagine =). Would that be a sufficient amount of data to understand which users of "swimming.com" are similar to one another? Maybe? Perhaps you could attempt to train a neural network based on what articles someone clicks on? People who read similar articles are similar to one another?

There is another problem though, which is that "swimming.com" would likely be a very niche website with a small number of visitors. The proposal I put up works by submitting an "embedding vector" to the browser when someone visits the publisher site (in this case "swimming.com"). So only people who had previously visited "swimming.com" would have an embedding vector sent to their browser.

 > As the owner of swimming.com, it sounds like I need to be prepared to produce the embedding vector myself, or hire someone to do it?

Yes.

> as the owner of swimming.com do I need to convince advertisers like TYR & Roka to deploy my JS for this to work?

Yes. So TYR and Roka would need to deploy some 3rd party JS on their website. When those people make a purchase, IF they had previously visited "swimming.com" and sent an "embedding vector" to the browser, these embedding vectors could be aggregated in some capacity. This would tell you something about which of the users of "swimming.com" make purchases on the TYR and Roka websites.

> after this occurs, as a publisher what specifically do I have? Do I have a vector id, or do I know the user browsing my site is interested in x, y, or z?

Once this aggregate is complete, you would have a vector like <24.5, -98.1, 15.2, -35.6, ... >

How would this be useful? It would only be useful for showing ads on "swimming.com". All of the users of "swimming.com" would have an embedding vector assigned to them, which could be compared with this aggregate learning about the customers of the TYR and Roka websites. In this way, "Swimming.com" could understand: "Ah! This reader of swimming.com seems to be similar to the set of other users of "swimming.com" who have previously made purchases on TYR's website. I guess I'll show them ads for TYR".

This example seems pretty far-fetched though. I hope this explanation makes this proposal more clear. 

What should be clear is that this proposal is *not* designed for ad-exchanges. Ad-exchanges have essentially zero first-party data about people. They only have 3rd party data. In a future where it's not possible to link together browsing activity across multiple websites, most ad-networks will have essentially nothing. People do not have a direct, first-party relationship with ad-networks. As such, ad-networks would seem to be unable to train an embedding.

**Comment by michaelkleber** [comment]
Hi folks: I think I'm more optimistic about this line of thinking than Ben sounds above.

I agree that looking at people's behavior on swimming.com alone is unlikely to be useful for guessing which lookalike audiences they should join: It might be a good way to generalize the membership criteria for a _small slice_ of interest groups, but would probably be useless for many others.

But if an ad-tech third party has presence on a lot of different publisher sites, perhaps they can repeat this process on each one of them, independently.  That is: For every 1p site that a 3p works with, the 3p can use the approach Ben describes to build a _per-1p_ model of which type of visitors to that 1p ought to be in each of that 3p's lookalike audiences.

I understand the appeal of making global cross-site user models for this purpose, as suggested in https://github.com/MagniteEngineering/Gatekeeper, but of course that is much harder from the privacy point of view.  If a large number of single-site models can do almost as good a job as one cross-site model, it would be vastly easier to incorporate into a post-tracking world.

**Comment by jdwieland8282** [comment]
thanks @benjaminsavage your comments are mostly inline with what I was expecting. 

> There is no "threshold". You would simply need enough information to be capable of sensibly assigning an "embedding vector" to each browser instance.

This is one of the red flags I recall from your original presentation of this proposal. A company like FB has lots of pageviews/FPD, swimming.com does not, hence the desire to share FPD cross domain within defined privacy boundaries, agreed upon standards, & most important trust between a browser and a gatekeeper (audience assembler) 


> What should be clear is that this proposal is _not_ designed for ad-exchanges. Ad-exchanges have essentially zero first-party data about people.

agreed, however running an ad exchange is only part of our remit to pubs, we often act as a technology provider which expands and contracts our role over time. Pubs of all shapes and sizes are focused on content, and often delegate tasks like audience modeling to other tech vendors. 

Specifically we aren't proposing that exchanges "train an embedding", we are proposing that an independent trusted party, a Gatekeeper, do that.

**Comment by jdwieland8282** [comment]
Hi @michaelkleber,

I'm not sure how your suggestion would work, w/o keeping track of which sites a browser interacts with.  

> But if an ad-tech third party has presence on a lot of different publisher sites, perhaps they can repeat this process on each one of them, independently. That is: For every 1p site that a 3p works with, the 3p can use the approach Ben describes to build a _per-1p_ model of which type of visitors to that 1p ought to be in each of that 3p's lookalike audiences.

We do indeed have a presence on many publishers sites, but w/o a persistent id used to say Jeffs Browser went to site A,B,& C, thus he is not only interested in Swimming but rather Triathlon.  This also has the problem of limiting a domains ability to find its next 100 customers.  Effectively, this is cross domain retargeting, unless I am missing something?

**Comment by michaelkleber** [comment]
Right, I'm asking what fraction of the "Lookalike audience" building could be done _without_ the "went to site A,B,& C" type of information, and instead using information about behavior from only one site at a time.

**Comment by jdwieland8282** [comment]
None in this proposal as written

> There is no "threshold". You would simply need enough information to be capable of sensibly assigning an "embedding vector" to each browser instance. If you have very little information, the quality of the data will be poor and the "embedding vectors" assigned will likely carry very little meaning or value.

**Comment by brodrigu** [comment]
@michaelkleber 

"Lookalike audiences" are a set of users who behave similarly to users who have been identified to be receptive to a given form of advertising (let's call this the conversion). The most basic form of observable behavior on the web is a page view. Up to a point, the more page views that can be observed, the more valuable the lookalike audience can be. At minimum you need to correlate between the publisher page-view and the conversion.

There are a number of ways a page-view could be processed into a content-category or an interest-group but they all stem from "went to site x". Later, a conversion is observed which is basically "went to site y" (the advertisers site). 

Large publishers, like Google or Facebook (sorry to call you out), see lots of users and many many page-views (I assume, correct me if I am wrong here). If we consider Google as a very very large "Site A" (with many pageviews with lots of interests built in) and an advertiser conversion as "Site B" then perhaps only site A & B are needed. However, if we consider a much smaller local newspaper as "Site A", the value of "went to site A,& B" is too small and sites C or more are needed to build something meaningful.

tldr; excluding the "big fish", no fraction of the "Lookalike audience" building could be done without the "went to site A B,& C" type of information.

We are trying to explore ways to utilize this information in a more privacy preserving way.

**Comment by jdwieland8282** [comment]
@michaelkleber Based on your comments today in the WC3BG meeting, 1b. style audiences (aka cohorts) assembly outside of a browser controlled server is off the table?  Is that a fair characterization?

**Comment by michaelkleber** [comment]
@jdwieland8282 I was trying to express the same skepticism I did in our [previous discussion](https://github.com/jkarlin/floc/issues/20#issuecomment-673671088): a server that collects cross-site data tied to some sort of user ID seems like far too much privacy risk.

That doesn't mean that use-case 1b is off the table, it just means we should look for different ways to accomplish it — which is what @benjaminsavage's proposal here was all about in the first place.

**Comment by jdwieland8282** [comment]
and by "a server" you mean one controlled by an entity not the Browser.  Browsers are still permitted to ship cross domain user ids to a server, for purposes of cohort assembly and making sure the number of users in the cohort is sufficiently high. 

wrt, Ben's proposal, I believe we all arrived at the same conclusion, essentially that this will not work well for publishers with small data sets.

**Comment by jwrosewell** [comment]
First party data options are also not options for publishers with small data sets. See this [short article from Axios](https://www.axios.com/new-york-times-advertising-792b3cd6-4bdb-47c3-9817-36601211a79d.html) discussing New York Times first party solution.

> "This can only work because we have 6 million subscribers and millions more registered users that we can identify and because we have a breadth of content," says Allison Murphy, Senior Vice President of Ad Innovation.

> Murphy notes: "While a differentiator and I'm thrilled about it, this isn't a path available for every publishers, especially not local who don’t have the scale of resources for building from scratch."

**Comment by michaelkleber** [comment]
> Browsers are still permitted to ship cross domain user ids to a server, for purposes of cohort assembly and making sure the number of users in the cohort is sufficiently high.

@jdwieland8282 As we've discussed several times before (e.g. [here](https://github.com/jkarlin/floc/issues/20#issuecomment-673562620) and [here](https://github.com/jkarlin/floc/issues/18#issuecomment-678299188)), that is not the Chrome FLoC plan.  Regarding cohort assembly, the [original explainer](https://github.com/jkarlin/floc#overview) was very explicit about this:

> The central idea is that these input features to the [clustering] algorithm, including the web history, are kept local on the browser and are not uploaded elsewhere.

And for making sure cohorts are large enough, we can use infrastructure like a [Multi-Browser Aggregation Service](https://github.com/WICG/conversion-measurement-api/blob/master/SERVICE.md), which does not involve sending user ids to a server.
